lib/AST/DiagnosticConsumer.cpp

```diff
@@ -14,15 +14,148 @@
 //
 //===----------------------------------------------------------------------===//
 
-#define DEBUG_TYPE "swift-basic"
+#define DEBUG_TYPE "swift-ast"
 #include "swift/AST/DiagnosticConsumer.h"
 #include "swift/AST/DiagnosticEngine.h"
+#include "swift/Basic/Defer.h"
+#include "swift/Basic/SourceManager.h"
 #include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/StringSet.h"
 #include "llvm/Support/Debug.h"
 #include "llvm/Support/raw_ostream.h"
 using namespace swift;
 
-DiagnosticConsumer::~DiagnosticConsumer() { }
+DiagnosticConsumer::~DiagnosticConsumer() = default;
+
+llvm::SMLoc DiagnosticConsumer::getRawLoc(SourceLoc loc) {
+  return loc.Value;
+}
+
+FileSpecificDiagnosticConsumer::FileSpecificDiagnosticConsumer(
+    SmallVectorImpl<ConsumerPair> &consumers)
+  : SubConsumers(std::move(consumers)) {
+#ifndef NDEBUG
+  llvm::StringSet<> seenFiles;
+  for (const ConsumerPair &consumerPair : SubConsumers) {
+    if (consumerPair.first.empty())
```

Comment:
When would the first element ever be empty? Is it worth a comment somewhere?

Reply:
That's the "empty files represent no affinity" thing from the doc comment of the constructor.

Reply:
Hmm... Maybe reify that into a (static) function that returns a bool, and takes a `ConsumerPair` argument, and call the function `hasAffinity`.